### PR TITLE
Remove colors from error messages when redirected to a file

### PR DESCRIPTION
**Pull request by msedlace**
This change makes sure that no escape characters for terminal colorizations are present in error logs when standard error output is redirected to a file.

### CODE REVIEW

File: lib/exceptions.h [hunk 1]
```diff
@@ -54,27 +56,47 @@ class CompilerBug final : public P4CExceptionBase {
  public:
     template <typename... T>
     CompilerBug(const char* format, T... args)
-            : P4CExceptionBase(format, args...)
-    { message = cstring(ANSI_RED) + "Compiler Bug" + ANSI_CLR + ":\n" + message; }
+            : P4CExceptionBase(format, args...) {
+        // Check if output is redirected and if so, then don't color text so that
+        // escape characters are not present
+        const bool isCerrRedirected = ttyname(fileno(stderr)) == nullptr;  // NOLINT(runtime/threadsafe_fn)
```
Comment (mbudiu-vmw) [inline lib/exceptions.h:62]: Please move this code into a helper function, e.g., `colorize`.
You can also probably cache the result of the check.

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]

File: lib/exceptions.h (hunk 1 above)
Comment (msedlace) [reply lib/exceptions.h]: Sure thing, I moved it to its own function.

File: lib/exceptions.h [hunk 2]
```diff
@@ -29,6 +30,11 @@ constexpr char ANSI_RED[]  = "\e[31m";
 constexpr char ANSI_BLUE[] = "\e[34m";
 constexpr char ANSI_CLR[]  = "\e[0m";
 
+/// Checks if stderr is redirected to a file
+inline bool is_cerr_redirected() {
```
Comment (mbudiu-vmw) [inline lib/exceptions.h:34]: I would suggest having a static bool to save the result after the first call. It won't change later, so you can avoid calling ttyname. It is unlikely that error is called many times, but it is still healthier to save the value.

File: lib/exceptions.h [hunk 3]
```diff
@@ -54,27 +60,40 @@ class CompilerBug final : public P4CExceptionBase {
  public:
     template <typename... T>
     CompilerBug(const char* format, T... args)
-            : P4CExceptionBase(format, args...)
-    { message = cstring(ANSI_RED) + "Compiler Bug" + ANSI_CLR + ":\n" + message; }
+            : P4CExceptionBase(format, args...) {
+        // Check if output is redirected and if so, then don't color text so that
+        // escape characters are not present
+        message = (is_cerr_redirected() ? "" : cstring(ANSI_RED)) + "Compiler Bug"
```
Comment (mbudiu-vmw) [inline lib/exceptions.h:66]: I suggested moving the entire expression `(is_cerr_redirected() ? "" : cstring(ANSI_RED))` into a helper function. Perhaps colorize(ANSI_RED) and clear_colors().

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]

File: lib/exceptions.h [hunk 4]
```diff
@@ -31,8 +31,26 @@ constexpr char ANSI_BLUE[] = "\e[34m";
 constexpr char ANSI_CLR[]  = "\e[0m";
 
 /// Checks if stderr is redirected to a file
+/// Check is done only once and then saved to a static variable
 inline bool is_cerr_redirected() {
-    return ttyname(fileno(stderr)) == nullptr;  // NOLINT(runtime/threadsafe_fn)
+    static bool is_redir = ttyname(fileno(stderr)) == nullptr;  // NOLINT(runtime/threadsafe_fn)
```
Comment (mbudiu-vmw) [inline lib/exceptions.h:36]: you are not saving anything, the call to ttyname is still made every time.
you need to also remember whether is_redir has been initialized, using a second static variable.

File: lib/exceptions.h [hunk 5]
```diff
@@ -31,8 +31,26 @@ constexpr char ANSI_BLUE[] = "\e[34m";
...
+    return is_redir;
+}
+
+/// Function used to delete color in case of cerr output being redirected
+inline const char *cerr_colorize(const char *color) {
+    if(is_cerr_redirected()){
```
Comment (mbudiu-vmw) [inline lib/exceptions.h:42]: we leave spaces in ifs.

**Review by mbudiu-vmw — CHANGES_REQUESTED** [review]

**Review by mbudiu-vmw — APPROVED** [review]

**Pull request merged**